### DISCUSSION TITLE
why `http body` types are scattered throughout the crate

### DISCUSSION THREAD

**Discussion by ghost** [tokio-rs/axum#2836] [General]
`axum::Form`, `axum::Json`, `axum::response::Html`, `axum::extract::Multipart` are different http body types. So, I think they should be inside the `axum::body` module

**Comment by mladedav** [comment]
Notice that your examples include both `extract` and `response` structs. These serve different things, one is for requests, the other for responses.

And then I'd say it's better to have `FromRequest` extractors near `FromRequestParts` extractors than near `IntoResponse` implementations.

I'm not saying the code organization is perfect or that other stances wouldn't work, but I don't think we should simply move everything under `axum::body` which I think currently represents only the actual body, i.e. the bytes of the protocol and not the represented data.

If you don't mind me asking, what prompted this? Were you looking for something and was it hard to find the correct extractor/response type?

**Comment by ghost** [comment]
I completely get your point. But for a new comer, its quite daunting to figure out everything by reading docs only.

But in case of `axum::Json` both server and client can send that, so according to you it should be both in `axum::extract` and `axum::response` which actually does. But this creates multiple paths to a same type. The first thing that comes to mind when you hear types like `Json`, `Html`, `Multipart` etc. that its a `"body type"` then we try to figure out whether the client sends it or the server.

Keeping every `"body types"` under `axum::body` will easily implicate that `T` inside `Request<T>` and `Response<T>` will have types from `axum::body`.

And for those who are trying to find these types in the extract module for them `axum::extract` should have a soft link to those types.

I think that `axum` should be as simple as `http` crate. I think that libraries should be made in such a way that it can be memorized in the easiest way possible.

**Reply by jplatte** [reply]
Just because it would be easier for you, doesn't mean it would be easier for others.

Maybe the docs for the body module should be updated to link to the extract and response modules though, for people looking for these types there.
